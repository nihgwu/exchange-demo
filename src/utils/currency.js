export const currencies = {
  GBP: {
    code: 'GBP',
    symbol: '£',
    description: 'British Pound',
  },
  EUR: {
    code: 'EUR',
    symbol: '€',
    description: 'Euro',
  },
  USD: {
    code: 'USD',
    symbol: '$',
    description: 'American Dollar',
  },
};

export const currencyCodes = Object.keys(currencies);

export const getCodeIndex = code => currencyCodes.indexOf(code);

export const getNextCode = code => {
  const idx = getCodeIndex(code);
  let nextIdx = idx + 1;
  if (nextIdx >= currencyCodes.length) nextIdx = 0;

  return currencyCodes[nextIdx];
};

export const getCodeSymbol = code => {
  const currency = currencies[code];
  return currency ? currency.symbol : '';
};

// trim trailing zeros
export const trimNumber = (value, maxDigits = 2) => {
  return Number(value.toFixed(maxDigits));
};

export const formatMoney = (code, value) => {
  return `${getCodeSymbol(code)}${trimNumber(value)}`;
};

export const formatRate = (base, code, rates) => {
  if (!rates || !rates[code] || !rates[code]) return '';

  const rate = rates[code] / rates[base];
  const value = trimNumber(rate, 4);
  return `${getCodeSymbol(base)}1 = ${getCodeSymbol(code)}${value}`;
};
