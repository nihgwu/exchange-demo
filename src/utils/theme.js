import { createGlobalStyle } from 'styled-components';
import { normalize } from 'styled-normalize';

export const variant = (prop, variants) => props => variants[props[prop]];
export const boolean = (prop, style) => props => props[prop] && style;

export const GlobalStyle = createGlobalStyle`
  ${normalize};

  *,
  *:before,
  *:after {
    box-sizing: border-box;
  }
  
  body {
    margin: 0;
    padding: 0;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
      "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
      sans-serif;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
  }

  .slider {
    .paging-dot circle {
      fill: white !important;
    }

    .slider-list {
      margin: 0 !important;
    }
  }
`;

export const colors = {
  primary: '#3b7afa',
  primaryActive: '#1c4eb3',
  white: '#fff',
  light: 'rgba(255,255,255,0.6)',
};

export default {
  colors,
};
