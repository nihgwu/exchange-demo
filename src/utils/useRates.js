import { useRef, useState, useEffect } from 'react';

import { getRates, cancelGetRates } from 'services/rate';

const TIMEOUT = 10 * 1000;

const useRates = () => {
  const initial = useRef(true);
  const [data, setData] = useState({
    loading: true,
    rates: null,
    error: null,
  });

  useEffect(() => {
    let interval;
    if (initial.current) {
      const callback = async () => {
        // cancel pending request
        cancelGetRates();
        try {
          const rates = await getRates();
          // TODO: we could avoid triggering re-render if rates stays the same
          setData({
            loading: false,
            rates,
            error: null,
          });
        } catch (error) {
          setData({
            loading: false,
            rates: null,
            error,
          });

          // TODO: if failed to fetch rates, retry immediately
          // clearInterval(interval);
          // interval = setInterval(callback, TIMEOUT);
          // callback();
        }
      };

      interval = setInterval(callback, TIMEOUT);
      callback();
      initial.current = false;
    }

    return () => clearInterval(interval);
  }, [setData]);

  return data;
};

export default useRates;
