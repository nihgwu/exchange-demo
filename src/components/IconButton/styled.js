import styled, { css } from 'styled-components';

import { boolean } from 'utils/theme';

export const Wrapper = styled.button`
  background: transparent;
  border: none;
  padding: 0;
  display: flex;
  flex-direction: column;
  align-items: center;

  ${boolean(
    'disabled',
    css`
      opacity: 0.6;
    `
  )}
`;

export const Circle = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  background: #2e6cdd;
  width: ${props => props.size}px;
  height: ${props => props.size}px;
  border-radius: 50%;
  border: 1px solid ${props => props.theme.colors.light};
`;

export const Svg = styled.svg`
  color: ${props => props.theme.colors.white};
`;

export const Label = styled.label`
  color: ${props => props.theme.colors.light};
  font-size: 14px;
  margin-top: 10px;
`;
