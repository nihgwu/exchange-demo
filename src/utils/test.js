import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import theme from 'utils/theme';

const ThemeWrapper = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: ThemeWrapper, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
