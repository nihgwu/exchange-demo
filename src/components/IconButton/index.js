import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';

import { Wrapper, Circle, Svg, Label } from './styled';

const IconButton = forwardRef(function IconButton(
  { children, icon, iconSize = 48, disabled, ...rest },
  ref
) {
  return (
    <Wrapper {...rest} ref={ref} disabled={disabled}>
      <Circle size={iconSize}>
        <Svg as={icon} />
      </Circle>
      {children && <Label>{children}</Label>}
    </Wrapper>
  );
});

IconButton.propTypes = {
  children: PropTypes.string,
  icon: PropTypes.elementType.isRequired,
  iconSize: PropTypes.number,
  disabled: PropTypes.bool,
};

export default IconButton;
