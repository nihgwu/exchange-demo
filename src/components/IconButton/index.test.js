import React from 'react';
import { render, fireEvent } from 'utils/test';

import IconButton from './index';

describe('IconButton', () => {
  it('should renders correctly', () => {
    const { container } = render(<IconButton icon="svg" size={24} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should accept children as label', () => {
    const { getByText } = render(<IconButton icon="svg">Action</IconButton>);

    expect(getByText('Action')).toBeTruthy();
  });

  it('should not invoke click event when disabled', () => {
    const handleClick = jest.fn();
    const { container, rerender } = render(
      <IconButton icon="svg" onClick={handleClick}>
        Action
      </IconButton>
    );

    fireEvent.click(container.firstChild);
    expect(handleClick).toHaveBeenCalledTimes(1);

    rerender(
      <IconButton disabled icon="svg" onClick={handleClick}>
        Action
      </IconButton>
    );

    fireEvent.click(container.firstChild);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
