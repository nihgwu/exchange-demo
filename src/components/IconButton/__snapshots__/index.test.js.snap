// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton should renders correctly 1`] = `
.c0 {
  background: transparent;
  border: none;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #2e6cdd;
  width: 48px;
  height: 48px;
  border-radius: 50%;
  border: 1px solid rgba(255,255,255,0.6);
}

.c2 {
  color: #fff;
}

<button
  class="c0"
  size="24"
>
  <div
    class="c1"
    size="48"
  >
    <svg
      class="c2"
    />
  </div>
</button>
`;
